File: framework/languages/dotnet/samples/TicTacToe/Server/Play/Adapters/ZLink/Spots/Handlers/PlayActorObserveMilestoneHandler.cs
using TicTacToe.Server.Play.Adapters.ZLink.Actors;
using TicTacToe.Shared.Contracts;
using Zlink.Framework.Contracts.Handlers;
using Zlink.Framework.Contracts.Spots;

namespace TicTacToe.Server.Play.Adapters.ZLink.Spots.Handlers;

internal sealed class PlayActorObserveMilestoneHandler(
    ILogger<PlayActorObserveMilestoneHandler> logger)
    : IZLinkEntrySpotActorRequestHandler<PlayEntrySpot, PlayActor, ObserveMilestoneReq, ObserveMilestoneRes>
{
    public async ValueTask<ObserveMilestoneRes> HandleAsync(
        PlayEntrySpot entrySpot,
        PlayActor actor,
        ZLinkSpotActorRequestContext context,
        ObserveMilestoneReq message,
        CancellationToken cancellationToken)
    {
        _ = context;
        _ = message;

        logger.LogInformation(
            "actor: ObserveMilestoneReq received. actor={ActorId}",
            actor.ActorId);

        await entrySpot.SubscribeMilestoneAsync(actor, cancellationToken);

        logger.LogInformation(
            "actor: ObserveMilestoneReq completed. actor={ActorId}, subscribed={Subscribed}",
            actor.ActorId,
            true);
        return new ObserveMilestoneRes(true);
    }
}

File: framework/languages/node/samples/TicTacToe.Ts/Server/Play/Infrastructure/ZLink/Spots/EntrySpot/Handlers/play-actor-observe-milestone-handler.ts
import { Injectable } from '@nestjs/common';
import { ZLinkSpotActorRequest } from '@zlink-systems/framework';
import { observeMilestoneRes, PacketNames } from '../../../../../../../Shared/Contracts/messages';
import type {
  ZLinkSpotActorRequestContext
} from '@zlink-systems/framework';
import type {
  ObserveMilestoneReq,
  ObserveMilestoneRes,
} from '../../../../../../../Shared/Contracts/messages';
import type { PlayActor } from '../../../Actors/play-actor';
import { MilestoneObserverRegistry } from '../play-entry-spot';

@Injectable()
class PlayActorObserveMilestoneHandler {
  constructor(private readonly observers: MilestoneObserverRegistry) {}

  @ZLinkSpotActorRequest(PacketNames.observeMilestoneReq)
  async handle(
    actor: PlayActor,
    context: ZLinkSpotActorRequestContext,
    _request: ObserveMilestoneReq
  ): Promise<ObserveMilestoneRes> {
    void context;
    this.observers.subscribe(actor.actorId);
    return observeMilestoneRes(true);
  }
}

export { PlayActorObserveMilestoneHandler };
